### PR TITLE
[Wallet/iOS] Fix #35988: Deposit Support on Wallet iOS

### PR DESCRIPTION
**Pull request by nuo-xu**
Resolves https://github.com/brave/brave-browser/issues/35988

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Screens that support deposit:
1. Wallet Panel 
2. Portfolio 
3. Asset Details (from regular token & from market tab)
4. Accounts tab (each account's vertical ... button)
5. Account Details

### **4 flows of deposit when user has account**

**1. No pre-filled token or pre-filled account (Wallet Panel, Portfolio)
For this flow the test steps:**

1. tap `Deposit` button,
2. You will see a list of all available token for deposit (should be all tokens from token registry and user assets. NFTs are excluded since they are not considered as depositable)
3. You will have a network filter at the bottom left 
4. tap filter button and filter by some network and save it
5. Observe the token list should be updated by the network filter 
6. There is also a search filed
7. Type some query to see of the token list gets updated by the input
8. tap any token 
9. it will bring you to the deposit screen
10. observe you see the correct account name, account address and the QR code
11. observe you will see a bottom disclosure if the token you picked in the previous screen is `.eth` coin type
12. in this flow, you will see an account picker with a default account selected
13. click this picker it will give you all accounts that has the same coin type as the token you picked in the previous screen. 
14. pick a new account in the account selection screen
15. it will dismiss the screen and back to the deposit screen with updated UI

**2. Has a pre-filled token but no pre-filled account (regular token Asset Details)**

1. tap `Deposit` button,
2.  it will bring you to the deposit screen directly since we already know the token coin type.
3.  observe you see the correct account name, account address and the QR code
4.  observe you will see a bottom disclosure if the token you picked in the previous screen is `.eth` coin type
5.  in this flow, you will see an account picker with a default account selected
6.  click this picker it will give you all accounts that has the same coin type as the token coming from the asset details or the market tab 
7.  pick a new account in the account selection screen
8.  it will dismiss the screen and back to the deposit screen with updated UI

**3. Has a pre-filled query, but no pre-filled token or pre-filled account (coin market Asset Details)**

1. tap `Deposit` button,
2. You will see a list of all available token for deposit, those tokens are either token symbol case insensitive contains coin market's symbol or token name case insensitive contains coin market's symbol.
3. tap any token
4.  it will bring you to the deposit screen directly since we already know the token coin type.
5.  observe you see the correct account name, account address and the QR code
6.  observe you will see a bottom disclosure if the token you picked in the previous screen is `.eth` coin type
7.  in this flow, you will see an account picker with a default account selected
8.  click this picker it will give you all accounts that has the same coin type as the token coming from the asset details or the market tab 
9.  pick a new account in the account selection screen
10.  it will dismiss the screen and back to the deposit screen with updated UI

**4. Has a pre-filled account (account tabs, account details)**

1. tap `Deposit` button,
2. it will bring you to the deposit screen directly since we already know the account coin type
3. .observe you see the correct account name, account address and the QR code
4.  observe you will see a bottom disclosure if the token you picked in the previous screen is `.eth` coin type
5.  in this flow, you wont have an account picker since we have a pre-filled account information. 

### No Account Case
1. create a new wallet which will automatically give an eth account and a sol account but NO filecoin account
2. tap deposit in Portfolio 
3. tap filecoin on mainnet 
4. a account creation popup will show up
5. tap no will dismiss the popup
6. tap filecoin on mainnet again
7. tap ok to create a filecoin account 
8. the account creation screen will be dismissed after the account has been created
9. and user then will be redirect to the deposit screen 
10. check if the screen is displaying correct address information and qr code

1. create a new wallet which will automatically give an eth account and a sol account but NO filecoin account
2.  tap filecoin on mainnet in asset list or any other places that will bring you to the asset details screen
3.  tap deposit
4.  a account creation popup will show up
5.  tap no will dismiss the popup
6.  tap filecoin on mainnet again
7.  tap ok to create a filecoin account 
8.  the account creation screen will be dismissed after the account has been created
9.  and user then will be redirect to the deposit screen 
10.  check if the screen is displaying correct address information and qr code

Note: In this PR, we also updated and refactored `TokenList`, so it is required to do some sanity test in asset search in main wallet (magnifyingglass button), asset search in Buy, asset search in Swap to see if displaying and selection is still working as expected.

### CODE REVIEW

**Review by StephenHeaps — CHANGES_REQUESTED** [review]
Resolved: ~The new Deposit view isn't loading tokens or networks on iOS 15 (iPhone & iPad, seems iOS 15 related):~
iOS 15 | iOS 15 - Deposit Network Filters
--|--

File: ios/brave-ios/Sources/BraveWallet/Crypto/BuySendSwapDeposit/DepositTokenView.swift [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+                      String.localizedStringWithFormat(
+                        Strings.Wallet.userAssetSymbolNetworkDesc,
+                        assetViewModel.token.symbol,
+                        assetViewModel.network.chainName
+                      )
+                    )
+                    .font(.caption)
+                    .foregroundColor(Color(.braveLabel))
+                  }
+                }
```
[the remark this answers is not in the document]
Comment (StephenHeaps) [reply ios/brave-ios/Sources/BraveWallet/Crypto/BuySendSwapDeposit/DepositTokenView.swift:76]: I think we could actually re-use `TokenList` here, and take the opportunity to update it's UI to v2 style at the same time? Would update UI to v2 for a few other views too 🙂

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+              !allUserTokens.contains(where: { $0.id == token.id })
+            },
+            sortOrder: assetsForNetwork.sortOrder
+          )
+        }
+
+      let selectedNetworks = networkFilters.filter(\.isSelected).map(\.model)
+      let filteredNetworkAssets = (allUserAssets + allBlockchainTokens).filter { networkAssets in
+        selectedNetworks.contains {
+          $0.chainId.lowercased() == networkAssets.network.chainId.lowercased()
```
[the remark this answers is not in the document]
Comment (nuo-xu) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift:122]: i will update caseInsensitiveCompare(_:). query is now handled inside `TokenList`

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        networkAssets.tokens
+          .sorted { $0.symbol < $1.symbol }
+          .compactMap { token in
+            if !query.isEmpty,  // only if we have a filter query
+              !(token.symbol.localizedCaseInsensitiveContains(query)
+                || token.name.localizedCaseInsensitiveContains(query))
+            {
+              // token does not match query
+              return nil
+            }
```
[the remark this answers is not in the document]
Comment (nuo-xu) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift:135]: all those views you mentioned is now using `TokenList` which will handle search query.

File: ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            if !token.isErc721 && !token.isNft {  // no deposit for NFTs
+              return AssetViewModel(
+                groupType: .none,
+                token: token,
+                network: networkAssets.network,
+                price: "",
+                history: [],
+                balanceForAccounts: [:]
+              )
```
[the remark this answers is not in the document]
Comment (nuo-xu) [reply ios/brave-ios/Sources/BraveWallet/Crypto/Stores/DepositTokenStore.swift:144]: i can reduce it as a tuple.

**Review by StephenHeaps — COMMENTED** [review]
UI looks great. I love that we were able to re-use & update `TokenList` to get v2 UI styling for Search Assets and Buy & Swap asset selection as well.

One last issue to fixup; selecting Filecoin Mainnet asset from Deposit assets list without a Filecoin account created:


Filecoin on Filecoin Mainnet w:o Filecoin account

**Review by StephenHeaps — APPROVED** [review]

**Comment by srirambv** [comment]
Verification passed on iPhone 13 with iOS 17.4.1 running 1.66 (59)
- Verified steps from PR description
- Verified Deposit works as expected 
- Verified Deposit option works from Portfolio, Account details menu, Asset details screen and Market token details screen
- Verified Deposit auto fills token details when opened from Asset details screen
- Verified Deposit screen for Eth tokens shows disclaimer message
- Verified searching for a token on multichain the disclaimer message is only shown for ETH/EVM network tokens
- Verified when there is no account created, selecting deposit brings up the account creation message. Additional notes [here](https://github.com/brave/brave-core/pull/22816#issuecomment-2037620975)

https://github.com/brave/brave-core/assets/17010094/0bf04b0c-1945-40d3-bde4-df1e9b020276

**Pull request merged**